CRM/Core/Page/File.php

```diff
@@ -42,15 +42,24 @@ public function run() {
     $download = CRM_Utils_Request::retrieve('download', 'Integer', $this, FALSE, 1);
     $disposition = $download == 0 ? 'inline' : 'download';
 
-    $entityId = CRM_Utils_Request::retrieve('eid', 'Positive', $this, TRUE); // Entity ID (e.g. Contact ID)
+    $entityId = CRM_Utils_Request::retrieve('eid', 'Positive', $this, FALSE); // Entity ID (e.g. Contact ID)
     $fieldId = CRM_Utils_Request::retrieve('fid', 'Positive', $this, FALSE); // Field ID
-    $fileId = CRM_Utils_Request::retrieve('id', 'Positive', $this, TRUE); // File ID
+    $fileId = CRM_Utils_Request::retrieve('id', 'Positive', $this, FALSE); // File ID
+    $fileName = CRM_Utils_Request::retrieve('filename', 'String', $this, FALSE);
+    if (empty($fileName) && (empty($entityId) || empty($fileId))) {
+      CRM_Core_Error::statusBounce("Either FIlename or Entity ID and FIle Id need to be passed in to retrieve files");
+    }
     $hash = CRM_Utils_Request::retrieve('fcs', 'Alphanumeric', $this);
-    if (!CRM_Core_BAO_File::validateFileHash($hash, $entityId, $fileId)) {
+    if (!CRM_Core_BAO_File::validateFileHash($hash, $entityId, $fileId) && empty($fileName)) {
       CRM_Core_Error::statusBounce('URL for file is not valid');
     }
-
-    list($path, $mimeType) = CRM_Core_BAO_File::path($fileId, $entityId);
+    if (!empty($fileName)) {
+      $mimeType = '';
+      $path = CRM_Core_Config::singleton()->customFileUploadDir . $fileName;
```

Comment:
I think this needs a bit more care to avoid `../` trickery?

Reply:
@totten this was exactly as it was before see https://github.com/civicrm/civicrm-core/commit/2762985c11e01ede473d86a33af279bc341f9a46#diff-b4dd4746dff586ca84ff742ac60afdacL42

Reply:
Good thing we removed it! 😜

More seriously... i've seen several formulations that effectively prohibit `..`, although they differ a bit on what files and subfolders are allowed. Examples:

* Assert that `$fileName` does not (a) begin with a dot or (b) include any `/` or any `\`. (This prohibits all subfolders, parent folders, and dot files -- only regular files stored directly in the data folder.)
* Assert that `$fileName` matches `basename($fileName)`. (This also only allows regular files from the regular data folder.)
* Assert that `$fileName` matches a stricter affirmative pattern (like "alphanumerics plus one dot plus 3-5 alphanumerics). But we don't have a large organic corpus of examples or a valid formal spec, so it can be hard to make a good stricter pattern.

Note: Take care that Unix systems are consistent about using `/`, but PHP-Windows can be inconsistent about `\` and `/`, so a more defensive approach is to check both.

Reply:
@seamuslee001 I pushed up a change with a `basename()` trick.